kingpin/actors/aws/ecs.py

```diff
@@ -181,19 +174,113 @@
+    def _register_task(self, task_definition):
+        """Registers a task on ECS.
+
+        Args:
+            task_definition: dict of ECS task definition parameters.
+
+        Returns:
+            tuple: ('family', 'revision', 'task_definition_name').
+        """
+
+        family = task_definition['family']
+
+        self.log.info('Registering task definition with family {}'.format(
+            family))
+        try:
+            response = yield self.thread(
+                self.ecs_conn.register_task_definition, **task_definition)
+        except Boto3Error as e:
+            raise exceptions.UnrecoverableActorFailure(e.message)
+
+        # Parse data from the server's response.
+        task_definition = response['taskDefinition']
+        family = task_definition['family']
+        revision = task_definition['revision']
+
+        task_definition_name = '{}:{}'.format(family, revision)
+        self.log.info('Task definition {} registered'.format(
+            task_definition_name))
+        raise gen.Return((family, revision, task_definition_name))
+
+    @staticmethod
+    def _load_task_definition(task_definition_file, tokens):
+        """Loads and verifies a task definition template file, and interpolates
+        tokens.
+
+        Args:
+            task_definition_file: task definition file to load.
+            tokens: dict of key/value pairs to interpolate into the file.
+
+        Returns:
+            Resulting task definition dict.
+        """
+        task_definition = utils.convert_script_to_dict(
+            task_definition_file, tokens)
+        try:
+            jsonschema.validate(task_definition,
+                                TASK_DEFINITION_SCHEMA)
+        except jsonschema.exceptions.ValidationError as e:
+            raise exceptions.InvalidOptions(e)
+        return task_definition
+
+    @staticmethod
+    def _load_service_definition(service_definition_file, tokens):
+        """Loads and verifies a service definition template file, and interpolates
+        tokens. The service definition template file can be None.
+
+        Args:
+            service_definition_file: service definition file to load.
+                If None or an empty string, this returns only defaults.
+            tokens: dict of key/value pairs to interpolate into the file.
+
+        Returns:
+            Resulting service definition dict.
+        """
+        if not service_definition_file:
+            service_definition = {}
+        else:
+            service_definition = utils.convert_script_to_dict(
+                service_definition_file, tokens)
+            try:
+                jsonschema.validate(service_definition,
+                                    SERVICE_DEFINITION_SCHEMA)
+
+            except jsonschema.exceptions.ValidationError as e:
+                raise exceptions.InvalidOptions(e)
+
+        # Set default values.
+        service_definition.setdefault(
```

Comment:
dict.setdefault(KEY, VALUE) will set the value of KEY to VALUE iff KEY is not in the dict yet.